File: api/src/main/java/mb/pie/api/Layer.java
package mb.pie.api;

import java.io.Serializable;

/**
 * Internal layer for intercepting parts of task execution, used for validation.
 */
public interface Layer {
    void requireTopDownStart(TaskKey key, Serializable input);

    void requireTopDownEnd(TaskKey key);

    void validateVisited(TaskKey key, Task<?> task, TaskData data);

    void validatePreWrite(TaskKey key, TaskData data, StoreReadTxn txn);

    void validatePostWrite(TaskKey key, TaskData data, StoreReadTxn txn);
}

File: api/src/main/kotlin/mb/pie/api/Layer.kt
package mb.pie.api

/**
 * Internal layer for intercepting parts of task execution, used for validation.
 */
interface Layer {
  fun requireTopDownStart(key: TaskKey, input: In)
  fun requireTopDownEnd(key: TaskKey)
  fun <I : In, O : Out> validatePreWrite(key: TaskKey, data: TaskData<I, O>, txn: StoreReadTxn)
  fun <I : In, O : Out> validatePostWrite(key: TaskKey, data: TaskData<I, O>, txn: StoreReadTxn)
}
